receiver/signalfxreceiver/receiver.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signalfxreceiver
+
+import (
+	"compress/gzip"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/gorilla/mux"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/observability"
+	"github.com/open-telemetry/opentelemetry-collector/oterr"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	sfxpb "github.com/signalfx/com_signalfx_metrics_protobuf"
+	"go.opencensus.io/trace"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultServerTimeout = 20 * time.Second
+
+	responseInvalidMethod      = "Only \"POST\" method is supported"
+	responseInvalidContentType = "\"Content-Type\" must be \"application/x-protobuf\""
+	responseInvalidEncoding    = "\"Content-Encoding\" must be \"gzip\" or empty"
+	responseErrGzipReader      = "Error on gzip body"
+	responseErrReadBody        = "Failed to read message body"
+	responseErrUnmarshalBody   = "Failed to unmarshal message body"
+	responseErrNextConsumer    = "Internal Server Error"
+
+	// Centralizing some HTTP and related string constants.
+	protobufContentType       = "application/x-protobuf"
+	gzipEncoding              = "gzip"
+	httpContentTypeHeader     = "Content-Type"
+	httpContentEncodingHeader = "Content-Encoding"
+)
+
+var (
+	errNilNextConsumer = errors.New("nil nextConsumer")
+	errEmptyEndpoint   = errors.New("empty endpoint")
+)
+
+// sfxReceiver implements the receiver.MetricsReceiver for SignalFx metric protocol.
+type sfxReceiver struct {
+	sync.Mutex
+	logger       *zap.Logger
+	config       *Config
+	nextConsumer consumer.MetricsConsumer
+	server       *http.Server
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+var _ receiver.MetricsReceiver = (*sfxReceiver)(nil)
+
+func (r *sfxReceiver) MetricsSource() string {
+	const metricsSource string = "SignalFx"
```

Comment:
Is there a reason for declaring this inside the func?

Reply:
Actually no :) will remove